src/main/java/de/zalando/aruha/nakadi/enrichment/MetadataEnrichmentStrategy.java

```diff
@@ -0,0 +1,39 @@
+package de.zalando.aruha.nakadi.enrichment;
+
+import de.zalando.aruha.nakadi.domain.EventType;
+import de.zalando.aruha.nakadi.exceptions.EnrichmentException;
+import de.zalando.aruha.nakadi.util.FlowIdUtils;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class MetadataEnrichmentStrategy implements EnrichmentStrategy {
+    @Override
+    public JSONObject enrich(final JSONObject event, final EventType eventType) throws EnrichmentException {
+        try {
+            final JSONObject metadata = event.getJSONObject("metadata");
+
+            setReceivedAt(metadata);
+            setEventTypeName(metadata, eventType);
+            setFlowId(metadata);
+        } catch (final JSONException e) {
+            throw new EnrichmentException("enrichment error", e);
+        }
+
+        return event;
+    }
+
+    private void setFlowId(final JSONObject metadata) {
+        metadata.put("flow_id", FlowIdUtils.peek());
```

Comment:
That's is good question. I did some research and discovered that it's implemented on top of log4j NDC. Reading NDC doc it states that it's isolated on a thread level. Since the thread that is executing this line is the request's thread, we are guaranteed to get the correct flow id.